File: 프로그래머스_JAVA/l2_조이스틱.java
import java.util.Arrays;
import java.util.stream.IntStream;

public class l2_조이스틱 {

    public static void main(String[] args){
        String testName = "JAZ";
        int testAnswer = solution(testName);
    }
    public static int solution(String name) {
        char[] getName = name.toCharArray();
        int[] target = new int[getName.length];
        for(int i =0; i<target.length; i++){
            target[i] = Math.min(getName[i] -'A', 'Z'-getName[i]+1);
        }
        System.out.println(Arrays.toString(target));
        int nowIndex = 0;
        int count = 0;
        while(true){
//        while (IntStream.of(target).sum()!=0){
//            System.out.println(nowIndex + " "+Arrays.toString(target)+" "+count);
            if (target[nowIndex]!=0){
                count += target[nowIndex];
                target[nowIndex] = 0;
            }
            if (IntStream.of(target).sum()==0){
                break;
            }

                int leftCnt =0, rightCnt = 0;
                for(int i=0; i<target.length; i++){
                    leftCnt++;
                    if(target[(nowIndex-leftCnt+target.length)%target.length]!=0) break;
                }
                for(int i=0; i<target.length;i++){
                    rightCnt++;
                    if(target[(nowIndex+rightCnt+target.length)%target.length]!=0) break;
                }

                if (leftCnt<rightCnt){
                    nowIndex = (nowIndex-leftCnt+target.length)%target.length;
                    count += leftCnt;
                } else{
                    nowIndex = (nowIndex+rightCnt+target.length)%target.length;
                    count += rightCnt;
                }



        }
        System.out.println(count);
        return count;
    }
}

File: 프로그래머스/Python/l2_조이스틱.py
def solution(name):
    target = [min(ord(n)-ord('A'), ord('Z') - ord(n)+1) for n in name]
    length = len(name)

    cnt = 0
    i = 0

    while True:
        # 1. move
        if target[i]!=0:
            cnt += target[i]
            target[i] = 0
        if sum(target) == 0:
            break

        # 2. left or right
        left_cnt = 0
        right_cnt = 0
        
        for l in range(length):
            if target[(i-l)%length]==0:
                left_cnt +=1
            else:
                break
        for r in range(length):
            if target[(i+r)%length]==0:
                right_cnt +=1
            else:
                break
        
        if left_cnt < right_cnt:
            cnt += left_cnt
            i = (i-left_cnt)%length
        else:
            cnt += right_cnt
            i = (i+right_cnt)%length

    return cnt


solution('JEROEN')
solution('JAN')
# def solution(name):
#     target = [min(ord(n)-ord('A'), ord('Z')-ord(n)+1)  for n in name]
#     length = len(name)
    
#     cnt = 0
#     i = 0
#     while True:
#         #1. move
#         if target[i]!=0:
#             cnt += target[i]
#             target[i] = 0
#         if sum(target)==0:
#             break

#         #2. < or >
#         left_cnt, right_cnt = 0,0
#         for l in range(length):
#             if target[(i-l)%length]==0:
#                 left_cnt +=1
#             else:
#                 break
#         for r in range(length):
#             if target[(i+r)%length]==0:
#                 right_cnt+=1
#             else:
#                 break
        
#         if left_cnt < right_cnt:
#             i = (i-left_cnt)%length
#             cnt+=left_cnt
#         else:
#             i = (i+right_cnt)%length
#             cnt+=right_cnt


            
#     return cnt